TABLESWITCH throw ArrayIndexOutOfBoundsException when the high value is Integer.MAX_VALUE

the source code will cause ArrayIndexOutOfBoundsException if the `high` value of `tableswitch` is Integer.MAX_VALUE：

```java
	/**
	 * Read needed data (e.g. index) from file.
	 */
	public TABLESWITCH(ByteSequence bytes) throws IOException {
		super(Constants.TABLESWITCH, bytes);

		int low = bytes.readInt();
		int high = bytes.readInt();

		matchLength = high - low + 1;
		fixedLength = (short) (13 + matchLength * 4);
		length = (short) (fixedLength + padding);

		match = new int[matchLength];
		indices = new int[matchLength];
		targets = new InstructionHandle[matchLength];

// if high=Integer.MAX_VALUE, it cause java.lang.ArrayIndexOutOfBoundsException
		for (int i = low; i <= high; i++) {
			match[i - low] = i;
		}

		for (int i = 0; i < matchLength; i++) {
			indices[i] = bytes.readInt();
		}
	}
```
exception like this:
```text
java.lang.RuntimeException: Crashed whilst crashing with this exception: org.aspectj.apache.bcel.generic.ClassGenException: java.lang.ArrayIndexOutOfBoundsException: Index 3 out of bounds for length 3
        at org.aspectj.weaver.bcel.BcelWeaver.weave(BcelWeaver.java:1728)
        at org.aspectj.weaver.bcel.BcelWeaver.weaveWithoutDump(BcelWeaver.java:1651)
        at org.aspectj.weaver.bcel.BcelWeaver.weaveAndNotify(BcelWeaver.java:1418)
        at org.aspectj.weaver.bcel.BcelWeaver.weave(BcelWeaver.java:1192)
        at org.aspectj.ajdt.internal.compiler.AjPipeliningCompilerAdapter.weaveQueuedEntries(AjPipeliningCompilerAdapter.java:510)
        at org.aspectj.ajdt.internal.compiler.AjPipeliningCompilerAdapter.afterCompiling(AjPipeliningCompilerAdapter.java:374)
        at org.aspectj.ajdt.internal.compiler.CompilerAdapter.ajc$afterReturning$org_aspectj_ajdt_internal_compiler_CompilerAdapter$2$f9cc9ca0(CompilerAdapter.java:69)
        at org.aspectj.org.eclipse.jdt.internal.compiler.Compiler.compile(Compiler.java:428)
        at org.aspectj.ajdt.internal.core.builder.AjBuildManager.performCompilation(AjBuildManager.java:1096)
        at org.aspectj.ajdt.internal.core.builder.AjBuildManager.performBuild(AjBuildManager.java:275)
        at org.aspectj.ajdt.internal.core.builder.AjBuildManager.batchBuild(AjBuildManager.java:188)
        at org.aspectj.ajdt.ajc.AjdtCommand.doCommand(AjdtCommand.java:103)
        at org.aspectj.ajdt.ajc.AjdtCommand.runCommand(AjdtCommand.java:47)
        at org.aspectj.tools.ajc.Main.run(Main.java:374)
        at com.hujiang.gradle.plugin.android.aspectjx.internal.procedure.AJXTask.call(AJXTask.kt:137)
        at com.hujiang.gradle.plugin.android.aspectjx.internal.procedure.DoAspectProcedure.runAJXTask(DoAspectProcedure.kt:252)
        at com.hujiang.gradle.plugin.android.aspectjx.internal.procedure.DoAspectProcedure.processDirectoryInput(DoAspectProcedure.kt:158)
        at com.hujiang.gradle.plugin.android.aspectjx.internal.procedure.DoAspectProcedure.access$processDirectoryInput(DoAspectProcedure.kt:12)
        at com.hujiang.gradle.plugin.android.aspectjx.internal.procedure.DoAspectProcedure$process$1$1$1.invoke(DoAspectProcedure.kt:32)
        at com.hujiang.gradle.plugin.android.aspectjx.internal.procedure.DoAspectProcedure$process$1$1$1.invoke(DoAspectProcedure.kt:31)
        at com.hujiang.gradle.plugin.android.aspectjx.internal.concurrent.BatchTaskScheduler$schedule$1.call(BatchTaskScheduler.kt:36)
        at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
        at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
        at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
        at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
        at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: org.aspectj.apache.bcel.generic.ClassGenException: java.lang.ArrayIndexOutOfBoundsException: Index 3 out of bounds for length 3
        at org.aspectj.apache.bcel.generic.Instruction.readInstruction(Instruction.java:233)
        at org.aspectj.apache.bcel.generic.InstructionList.<init>(InstructionList.java:190)
        at org.aspectj.apache.bcel.generic.MethodGen.<init>(MethodGen.java:211)
        at org.aspectj.weaver.bcel.LazyMethodGen.initialize(LazyMethodGen.java:370)
        at org.aspectj.weaver.bcel.LazyMethodGen.getReturnType(LazyMethodGen.java:917)
        at org.aspectj.weaver.bcel.LazyMethodGen.toShortString(LazyMethodGen.java:560)
        at org.aspectj.weaver.bcel.LazyMethodGen.print(LazyMethodGen.java:601)
        at org.aspectj.weaver.bcel.LazyClassGen.printOne(LazyClassGen.java:918)
        at org.aspectj.weaver.bcel.LazyClassGen.print(LazyClassGen.java:877)
        at org.aspectj.weaver.bcel.LazyClassGen.toLongString(LazyClassGen.java:865)
        at org.aspectj.weaver.bcel.BcelWeaver.weave(BcelWeaver.java:1726)
        ... 27 more
```
You can reproduce the bug if the weaved source code contain this:
```java
    private static final int CASE_1 = Integer.MAX_VALUE;
    private static final int CASE_2 = Integer.MAX_VALUE - 1;
    private static final int CASE_3 = Integer.MAX_VALUE - 2;

    public String switchTest(int i) {
        String content = "";
        switch (i) {
            case CASE_1:
                content = "CASE_1";
                break;
            case CASE_2:
                content = "CASE_2";
                break;
            case CASE_3:
                content = "CASE_3";
                break;
            default:
                break;
        }
        return content;
    }
```
I fixed it like this:
```java
    /**
     * Read needed data (e.g. index) from file.
     *
     * @param bytes bytes
     * @throws IOException IOException
     */
    public TABLESWITCH(ByteSequence bytes) throws IOException {
        super(Constants.TABLESWITCH, bytes);

        int low = bytes.readInt();
        int high = bytes.readInt();

        matchLength = high - low + 1;
        fixedLength = (short) (13 + matchLength * 4);
        length = (short) (fixedLength + padding);

        match = new int[matchLength];
        indices = new int[matchLength];
        targets = new InstructionHandle[matchLength];

        // bugfix: case使用的值可能是Integer.MAX_VALUE，i++后值溢出变为负数导致数组越界，所以需要重新修改算法
//        for (int i = low; i <= high; i++) {
//            match[i - low] = i;
//        }
        for (int i = 0; i < matchLength; i++) {
            match[i] = low + i;
        }

        for (int i = 0; i < matchLength; i++) {
            indices[i] = bytes.readInt();
        }
    }
```

--- a/bcel-builder/src/main/java/org/aspectj/apache/bcel/generic/TABLESWITCH.java
+++ b/bcel-builder/src/main/java/org/aspectj/apache/bcel/generic/TABLESWITCH.java
@@ -124,8 +124,8 @@ public class TABLESWITCH extends InstructionSelect {
 		indices = new int[matchLength];
 		targets = new InstructionHandle[matchLength];
 
-		for (int i = low; i <= high; i++) {
-			match[i - low] = i;
+		for (int i = 0; i < matchLength; i++) {
+			match[i] = low + i;
 		}
 
 		for (int i = 0; i < matchLength; i++) {

--- a/tests/bugs1919/github_190/SwitchCaseWith_Integer_MAX_VALUE.java
+++ b/tests/bugs1919/github_190/SwitchCaseWith_Integer_MAX_VALUE.java
@@ -1,0 +1,20 @@
+public class SwitchCaseWith_Integer_MAX_VALUE {
+  public static void main(String[] args) {
+    System.out.println(switchTest(Integer.MAX_VALUE));
+  }
+
+  static String switchTest(int i) {
+    switch (i) {
+      case Integer.MAX_VALUE:
+        return "CASE_1";
+      default:
+        return "";
+    }
+  }
+}
+
+aspect MyAspect {
+  before() : execution(* switchTest(*)) {
+    System.out.println(thisJoinPoint);
+  }
+}

--- a/tests/src/test/java/org/aspectj/systemtest/ajc1919/Bugs1919Tests.java
+++ b/tests/src/test/java/org/aspectj/systemtest/ajc1919/Bugs1919Tests.java
@@ -19,6 +19,9 @@ public class Bugs1919Tests extends XMLBasedAjcTestCase {
     runTest("declare annotation with SOURCE retention");
   }
 
+  public void testsSwitchWith_Integer_MAX_VALUE() {
+    runTest("switch with Integer.MAX_VALUE case");
+  }
   public static Test suite() {
     return XMLBasedAjcTestCase.loadSuite(Bugs1919Tests.class);
   }

--- a/tests/src/test/resources/org/aspectj/systemtest/ajc1919/ajc1919.xml
+++ b/tests/src/test/resources/org/aspectj/systemtest/ajc1919/ajc1919.xml
@@ -172,4 +172,17 @@
 		</compile>
 	</ajc-test>
 
+	<!-- RuntimeException in BcelWeaver, see https://github.com/eclipse/org.aspectj/issues/190 -->
+	<ajc-test dir="bugs1919/github_190" vm="1.5" title="switch with Integer.MAX_VALUE case">
+		<compile files="SwitchCaseWith_Integer_MAX_VALUE.java" options="-1.5 -showWeaveInfo">
+			<message kind="weave" text="Join point 'method-execution(java.lang.String SwitchCaseWith_Integer_MAX_VALUE.switchTest(int))'"/>
+    </compile>
+		<run class="SwitchCaseWith_Integer_MAX_VALUE">
+			<stdout>
+				<line text="execution(String SwitchCaseWith_Integer_MAX_VALUE.switchTest(int))"/>
+				<line text="CASE_1"/>
+			</stdout>
+		</run>
+	</ajc-test>
+
 </suite>

--- a/tests/src/test/resources/org/aspectj/systemtest/ajc1919/ajc1919.xml
+++ b/tests/src/test/resources/org/aspectj/systemtest/ajc1919/ajc1919.xml
@@ -163,13 +163,13 @@
 	</ajc-test>
 
 	<!-- Weaver error when declaring annotation with SOURCE retention, see https://bugs.eclipse.org/bugs/show_bug.cgi?id=366085 -->
-	 <ajc-test dir="bugs1919/366085" vm="1.5" title="declare annotation with SOURCE retention">
-		 <compile files="Application.java DeclareAnnotationsAspect.aj ToString.java Marker.java" options="-1.5 -showWeaveInfo">
-			 <message kind="weave" text="'Application' (Application.java:1) is annotated with @Marker type annotation from 'DeclareAnnotationsAspect'"/>
-			 <message kind="weave" text="'public void Application.new(int)' (Application.java:4) is annotated with @Marker constructor annotation from 'DeclareAnnotationsAspect'"/>
-			 <message kind="weave" text="'public int Application.getNumber()' (Application.java:8) is annotated with @Marker method annotation from 'DeclareAnnotationsAspect'"/>
-			 <message kind="weave" text="'public int number' of type 'Application' (Application.java) is annotated with @Marker field annotation from 'DeclareAnnotationsAspect'"/>
-		 </compile>
-		</ajc-test>
+	<ajc-test dir="bugs1919/366085" vm="1.5" title="declare annotation with SOURCE retention">
+		<compile files="Application.java DeclareAnnotationsAspect.aj ToString.java Marker.java" options="-1.5 -showWeaveInfo">
+			<message kind="weave" text="'Application' (Application.java:1) is annotated with @Marker type annotation from 'DeclareAnnotationsAspect'"/>
+			<message kind="weave" text="'public void Application.new(int)' (Application.java:4) is annotated with @Marker constructor annotation from 'DeclareAnnotationsAspect'"/>
+			<message kind="weave" text="'public int Application.getNumber()' (Application.java:8) is annotated with @Marker method annotation from 'DeclareAnnotationsAspect'"/>
+			<message kind="weave" text="'public int number' of type 'Application' (Application.java) is annotated with @Marker field annotation from 'DeclareAnnotationsAspect'"/>
+		</compile>
+	</ajc-test>
 
 </suite>

